Oop exercise

Tedzury opened the pull request:
### OOP exersice

[Demo](https://tedzury.github.io/Tiny_JS_world/) |
[Code base](https://github.com/Tedzury/Tiny_JS_world)

The code is submitted in a dedicated feature branch.

Only code files are submitted.

Please, review.

Tedzury commented:
Good morning. For the first view in self-check, everything seems to be alright. Humbly asking for review :)

OleksiyRudenko commented on submissions/Oleksii_Drohachov/OOP-exercise/index.js:
```diff
@@ -0,0 +1,98 @@
+    this.saying = saying;
+    this.friends = friends;
+    this.propsArray = [
+      "species",
+      "name",
+      "gender",
+      "saying",
+      "friends",
+      "legs",
```
A base class should never handle child classes' properties. This breaks Encapsulation, Open-Closed and Separation of Concerns principles as any changes to child classes would require changes in one (or more) base class up the hierarchy.

Look into method overloading to have consistent interface and user experience where method implementation depends on class yet allows to re-use parent class' method for a similar purpose.

OleksiyRudenko commented on submissions/Oleksii_Drohachov/OOP-exercise/index.js:
```diff
@@ -0,0 +1,98 @@
+  constructor(name, gender, saying, friends) {
+    super(name, gender, saying, friends);
+    this.species = "human";
+    this.legs = 2;
```
`legs` looks like a shared property. Should be generalized to a base class just like any other shared property.
Same refers to `species`.

The criteria of generalization is property per se. It can bear different values for different class instances.

OleksiyRudenko reviewed (changes requested):
@Tedzury good job!
Yet implementation needs to follow OOP and SOLID principles.

Tedzury commented:
Hello. Removed legs from base properties, added class Mammal and put legs to it.

OleksiyRudenko commented on submissions/Oleksii_Drohachov/OOP-exercise/index.js:
```diff
@@ -0,0 +1,116 @@
+          "species",
+          "gender",
+          "legs",
+          "hands",
```
Every class should handle the props it owns immediately.
As a developer you do not want to touch base classes when adding a new one. This is what encapsualtion and open-closed principles from OOP/SOLID are about.

OleksiyRudenko commented on submissions/Oleksii_Drohachov/OOP-exercise/index.js:
```diff
@@ -0,0 +1,116 @@
+
+const tinyWorldInhabitants = [barbos, sonya, oleksii, victoria, anjela];
+
+tinyWorldInhabitants.forEach((item) => print(item.output()));
```
`item` is a generic term.

OleksiyRudenko commented on submissions/Oleksii_Drohachov/OOP-exercise/index.js:
```diff
@@ -0,0 +1,116 @@
+          "saying",
+          "friends",
+        ])
+      : (this.prettyOutput = this.propsArray);
```
Nested ternaries considered a bad practice. Can we make this fragment more obvious? KISS

OleksiyRudenko reviewed (changes requested):
@Tedzury quite a decent job.
Let's make it more OOP/SOLID compliant.
Method overloading and extended `super` API are to the rescue.

Tedzury commented:
Good night. Slightly refactored output function. Removed propsArray, applied function overrading to output( ) method and modified it into child classes.

OleksiyRudenko commented on submissions/Oleksii_Drohachov/OOP-exercise/index.js:
```diff
@@ -0,0 +1,97 @@
+
+class Mammal extends Creature {
+  constructor(species, name, gender, saying, friends, legs) {
+    super(species, name, gender, saying, friends);git 
```
`git` here should have thrown an error. Didn't it?

OleksiyRudenko commented on submissions/Oleksii_Drohachov/OOP-exercise/index.js:
```diff
@@ -0,0 +1,97 @@
+  }
+
+  output() {
+    return `${this.name}; ${this.species}; ${this.gender}; ${this.saying}; ${this.friendsToString()}`;
```
More than 2 items. `Array#join` is still a thing.

Won't be a really good approach, but think of how destructuring could help you in more complicated cases:
`const { propA, propC, propB } = this`

OleksiyRudenko reviewed (changes requested):
@Tedzury well done.
Let's polish it using learnings from past tasks.

Tedzury commented:
Good day. Removed typo, apllied destructuring to properties in output( ) method.

OleksiyRudenko commented on submissions/Oleksii_Drohachov/OOP-exercise/index.js:
```diff
@@ -0,0 +1,99 @@
+  output() {
+    const { name, species, gender, saying } = this;
+
+    return `${name}; ${species}; ${gender}; ${saying}; ${this.friendsToString()}`;
```
Let's employ `Array#join`. Change this line only.
Remember, anything can be an array element. `[ func() ]` will add a result from func call as an array element.

OleksiyRudenko commented on submissions/Oleksii_Drohachov/OOP-exercise/index.js:
```diff
@@ -0,0 +1,99 @@
+  }
+}
+
+class Mammal extends Creature {
+  constructor(species, name, gender, saying, friends, legs) {
+    super(species, name, gender, saying, friends);
+    this.legs = legs;
+  }
+  output() {
+    return `${super.output()}; ${this.legs}`;
+  }
+}
```
`Mammal` is the only class that extends `Creature`. Why do we need `Mammal`?
KISS

OleksiyRudenko commented on submissions/Oleksii_Drohachov/OOP-exercise/index.js:
```diff
@@ -0,0 +1,99 @@
+  }
+}
+
+class Animal extends Mammal {
+  constructor(species, name, gender, saying, friends) {
+    super(species, name, gender, saying, friends, 4);
+  }
+}
```
`Animal` extends `Mammal` just to pass a specific value.
Solvable by setting default argument value.
KISS

OleksiyRudenko reviewed (changes requested):
@Tedzury let's optimize hierarchy.
Introducing classes just for the purpose of setting default values is an overkill.
I am not even sure we need classes for men and women. Or shy dogs and cats do not deserve sheDog/Cat and heDog/Cat classes?

Tedzury commented:
Good night. Removed unneded animal class, man and woman. Just passing legs as default value, and gender to humans as an argument. Also implemented array.join in output method

OleksiyRudenko commented on submissions/Oleksii_Drohachov/OOP-exercise/index.js:
```diff
@@ -0,0 +1,73 @@
+  output() {
+    const { name, species, gender, saying, friends } = this;
+
+    return `${name}; ${species}; ${gender}; ${saying}; ${friends.join(", ")}`;
```
There is still repetition (`"; "`). Pack whatever you need into an array and `Array#join`.
You will benefit a lot from recognizing use cases for `Array` methods.

OleksiyRudenko commented on submissions/Oleksii_Drohachov/OOP-exercise/index.js:
```diff
@@ -0,0 +1,73 @@
+  }
+}
+
+class Mammal extends Creature {
```
> `Mammal` is the only class that extends `Creature`. Why do we need `Mammal`?

Any good reason for this?

Also, obviously there is no good reason to prefer 4 for legs. Decisions should be explainable.

OleksiyRudenko reviewed (changes requested):
@Tedzury nearly there

Tedzury commented:
Good morning. Left only base class and Human/Dog/Cat, legs are passed as default property in child classes. Implemented array with props to use array.join in the output method. 

Sorry, but I feel very confused, I had exactly the same way of array.join in output two commits ago, but I undestood you in such way, that I shouldn't use it, so tried everything else, except manipulations with array :(

OleksiyRudenko commented on submissions/Oleksii_Drohachov/OOP-exercise/index.js:
```diff
@@ -0,0 +1,76 @@
+    this.saying = saying;
+    this.friends = friends;
+    this.legs = legs;
+    this.propsArray = [
```
Data type name as a part of variable name is not necessary (think if you switch to `Map` data type, you'd need to rename variable to reflect the change).
Plural noun is just sufficient.

Also `props` is quite specific term and kinda technical (as in e.g. "component props"). Here we mean "properties of inhabitant", just a normal word that in real life is never abbreviated. Do not save keystrokes on names, IDE will help you to autocomplete even German-language-style variable names.

OleksiyRudenko reviewed (approved):
@Tedzury great job done!
Make sure to put some learnings on your diary.
